datawig/simple_imputer.py

```diff
@@ -586,3 +603,65 @@ def load_hpo_model(self, hpo_name: int = None):
 
         self.imputer = Imputer.load(self.output_path)
 
+    def check_for_label_shift(self,
+                              target_data: pd.DataFrame) -> tuple:
+
+        """
+        Detect label shift in the validation data
+
+        :param test_data: data frame that contains labels
+        :param target_data: unlabelled data for which predictions are to be generated
+
+        :return: tuple of label weight dict
+        """
+
+        # labels need to be sorted consistently for computation of confusion matrices etc.
+        labels = sorted(self.imputer.label_encoders[0].idx_to_token.values())
+
+        # generate predictions for test and target data
+        imputed_target = self.predict(target_data)
+        # imputed_test = self.predict(test_data)
+
+        # invert normalized empirical confusion matrix for test data
+        ids = dict((label, index) for (index, label) in enumerate(labels))
+        confusion_matrix_from_file = self.load_metrics()['confusion_matrix']
+        confusion_test = np.empty([len(confusion_matrix_from_file), len(confusion_matrix_from_file)])
+
+        for row in confusion_matrix_from_file:
+            for entry in row:
+                if type(entry) is str:
+                    row_idx = ids[entry]
+                else:
+                    for col in entry:
+                        col_idx = ids[col[0]]
+                        confusion_test[row_idx, col_idx] = col[1]
+
+        confusion_test = confusion_test/confusion_test.sum()
+        confusion_inv_test = np.linalg.inv(confusion_test)
+
+        # compute estimates of label marginals for test and target data
+        # marginals_test = imputed_test[self.output_column+'_imputed'].value_counts(normalize=True, sort=False)[labels]
+        marginals_test = pd.Series(confusion_test.sum(axis=0), index=labels)
+        marginals_target = imputed_target[self.output_column+'_imputed'].value_counts(
+            normalize=True, sort=False)[labels]
+
+        # estimate the ratio of marginals and store as dictionary.
+        label_weights = confusion_inv_test.dot(marginals_target)
```

Comment:
why are we computing the inverse and not solving a linear system via QR or so?

Reply:
Thanks for catching that. We were interested in the inverse at some point, but now it's much better to use something more stable, PLU decomposition here. (np.linalg.solve)